### PR TITLE
gh-111225: Link extension modules against libpython on Android

### LINKED ISSUE #111225: Cross Compile for Android and issues with PyExc_OSError
# Bug report

### Bug description:

In trying to run Python 3.12 on Android. I have been cross-compiling Python for a number of years. Having built a 64bit version I am getting the following error when I try to run

I get the following error at runtime.
ImportError('dlopen failed: cannot locate symbol "PyExc_OSError" referenced by "/data/data/com.myapp/files/Python64/lib-dynload/_socket.cpython-312.so"...')

I have verified that the _socket.cpython-3.12.so exists.  I'm not really sure what other info would be beneficial for people to help me understand this error so I will start with this and happily add more details if requested.

### CPython versions tested on:

3.12

### Operating systems tested on:

Other

### Linked PRs
* gh-115780

### PR DESCRIPTION
**Pull request by mhsmith**
For the reasons explained in [PEP 738](https://peps.python.org/pep-0738/#linkage), extension modules must be linked against libpython on Android.

Cygwin has the same requirement, and it looks like there were two existing attempts to implement this:

* One of them, in makesetup, only applied to Cygwin. 
* The other, in configure.ac, applied to both Cygwin and Android, and worked by setting a LIBPYTHON variable. But this had no effect, because the variable wasn't actually being used anywhere.

I've removed the first attempt in favor of the second one, and fixed it as follows:

* Renamed the variable to MODULE_LDFLAGS to more accurately reflect its purpose.
* Edited makesetup to use the variable when linking extension modules.
* Edited the Makefile so that extension modules depend on libpython on these platforms.
* Restored `-fPIC` on Android. It was removed several years ago with a note that the toolchain used it automatically, but this is no longer the case. Omitting it causes all linker commands to fail with an error like `relocation R_AARCH64_ADR_PREL_PG_HI21 cannot be used against symbol '_Py_FalseStruct'; recompile with -fPIC`.

Before: 

```
$ readelf -d _socket.cpython-313.so | grep -E 'NEEDED|SONAME' 
 0x0000000000000001 (NEEDED)             Shared library: [libm.so]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so]
```
After:
```
 0x0000000000000001 (NEEDED)             Shared library: [libm.so]
 0x0000000000000001 (NEEDED)             Shared library: [libpython3.13.so.1.0]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so]
```

* Fixes #111225

### CODE REVIEW

File: Modules/makesetup [hunk 1]
```diff
@@ -87,18 +87,6 @@ esac
...
-case `uname -s` in
-CYGWIN*) if test $libdir = .
-	 then
-	 	ExtraLibDir=.
-	 else
-	 	ExtraLibDir='$(LIBPL)'
-	 fi
-	 ExtraLibs="-L$ExtraLibDir -lpython\$(LDVERSION)";;
-esac
-
```
Comment (erlend-aasland) [inline Modules/makesetup:101]: AFAIK, building on Cygwin has been broken for years. It is also not a supported platform. Removing cruft like this is ok.

**Review by erlend-aasland — DISMISSED** [review]
Looks good; thanks a lot for making this bug-fix in a separate PR.

**Comment by erlend-aasland** [comment]
Oops, seems this broke `--enable-shared` build on macOS:

```console
$ ./python.exe -m test
dyld[79567]: Library not loaded: /lib/libpython3.13d.dylib
  Referenced from: <91EB63F2-04C1-35AB-B557-218711027856> /Users/erlend.aasland/src/python/shared/python.exe
  Reason: tried: '/lib/libpython3.13d.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/lib/libpython3.13d.dylib' (no such file), '/lib/libpython3.13d.dylib' (no such file)
[1]    79567 abort      ./python.exe -m test
```

**Comment by erlend-aasland** [comment]
Hm, seems the macOS build is still broke for shared builds.

**Review by erlend-aasland — APPROVED** [review]

**Review by erlend-aasland — CHANGES_REQUESTED** [review]
Sorry for the back and forth; I had forgotten to clean the build folder. I redid my tests and I now can confirm that this breaks `--enable-shared` builds on macOS (`main` works fine).

**Comment by erlend-aasland** [comment]
> Sorry for the back and forth; I had forgotten to clean the build folder. I redid my tests and I now can confirm that this breaks `--enable-shared` builds on macOS (`main` works fine).

Aha, this PR does not include 074bbec9c4911da1d1155e56bd1693665800b814; rebasing onto `main` fixed the problem. Looks like this is good to go again 😄

**Review by erlend-aasland — APPROVED** [review]

**Comment by eli-schwartz** [comment]
> Renamed the variable to MODULE_LDFLAGS to more accurately reflect its purpose.

Renaming it breaks the de-facto API that the sysconfig module has provided for 5 years, and which downstream projects rely on.

> Cygwin has the same requirement, and it looks like there were two existing attempts to implement this:
> 
> * One of them, in makesetup, only applied to Cygwin.
> 
> * The other, in configure.ac, applied to both Cygwin and Android, and worked by setting a LIBPYTHON variable. But this had no effect, because the variable wasn't actually being used anywhere.

That check was originally written to configure the python-config script -- I later updated it to also configure the python.pc file. distutils had its own, second copy of the configure.ac logic which was probably dropped during the migration away from internally using distutils. Not really the fault of the variable itself but what can you do. :)

**Comment by mhsmith** [comment]
It's unfortunate that sysconfig exposes all the Makefile variables in this way, because most of them were created with no intention of becoming a stable API. Can you give some examples of projects that rely on this variable?

Setuptools, at least, does not rely on it. It has [its own logic](https://github.com/pypa/setuptools/blob/v72.1.0/setuptools/_distutils/command/build_ext.py#L747) for deciding which platforms need to link against libpython, and it already knows about Windows, Cygwin and Android.

**Comment by eli-schwartz** [comment]
https://github.com/mesonbuild/meson/blob/fa4f2339465ce3d755e2df802ebd5aa962e2ad27/mesonbuild/scripts/python_info.py#L66-L79

Lacking any formalization of the information which might come via #107956, it is necessary to poke at raw Makefile variables -- in this case, the raw Makefile variable that corresponds to whether the "C-API extension modules" pkg-config file is configured to link to libpython. The setuptools logic is actually a copy-paste of the stdlib distutils logic, which was *manually* kept in sync with Makefile.pre.in up until it got deleted from the stdlib.

**Comment by rgommers** [comment]
This looks like it's going to break the build for NumPy on Android - the flag to link to `libpython` is going to silently disappear. And there is a fair amount of usage of NumPy on Android, judging by bug reports over the past few years. And we won't be able to work around this for NumPy 2.1.0 which is about to be released with Python 3.13 support, even if Meson adapts to this rename.

`LIBPYTHON` is the most obvious way to determine whether the linker flag is needed yes or no. It's one of the more important flags one can access through `get_config_vars`. A `libpython` key is also present in the current draft of PEP 739: https://peps.python.org/pep-0739/#libpython for that reason. 

> It's unfortunate that sysconfig exposes all the Makefile variables in this way, because most of them were created with no intention of becoming a stable API

I couldn't agree more, but unfortunately that's the situation we're in. It's best not to rename variables unless the actual content/semantics of the variable changes. In this case, reverting the rename looks like the right thing to do.

**Comment by mhsmith** [comment]
Thanks for the explanation.

> This looks like it's going to break the build for NumPy on Android

Actually as @eli-schwartz says in https://github.com/python/cpython/issues/116745#issuecomment-2254636265, it looks like it will break the build on every platform *except* Android, because the Meson code defaults to linking against libpython when the variable is missing from sysconfig.

So I agree the rename must be reverted. I've done that in #122764.

**Comment by eli-schwartz** [comment]
Thanks for looking into this! :)

Fingers crossed that PEP 739 can be rolled out in the near future (hopefully as a beautiful gift included in 3.14).

**Comment by Sanny388cry** [comment]
mhsmith:android-link-modules

**Pull request merged**